frontends/ast/ast.cc

```diff
@@ -1100,6 +1095,12 @@ static AstModule* process_module(AstNode *ast, bool defer, AstNode *original_ast
 		ignoreThisSignalsInInitial = RTLIL::SigSpec();
 	}
 
+	for (auto &attr : ast->attributes) {
+		if (attr.second->type != AST_CONSTANT)
+			log_file_error(ast->filename, ast->linenum, "Attribute `%s' with non-constant value!\n", attr.first.c_str());
+		current_module->attributes[attr.first] = attr.second->asAttrConst();
+	}
+
```

Comment:
This will fail on attributes that have a compound expression as value instead of a literal.

```
(* foo = 1 + 2 *)
module test;
endmodule
```

```
$ ./yosys -Qp 'read_verilog -defer test.v'

-- Running command `read_verilog -defer test.v' --

1. Executing Verilog-2005 frontend: test.v
Parsing Verilog input from `test.v' to AST representation.
Storing AST representation for module `$abstract\test'.
test.v:2: ERROR: Attribute `\foo' with non-constant value!
```

Reply:
Hmm.. now it doesn't produce an error message if the expression is not constant. Maybe it would make sense to _only_ change the behavior for `top` attributes?

> The fact that it must be a constant expression should mean that it is evaluate-able even for deferred elaboration.

The whole idea of `-defer` is that we defer simplifying and evaluating the AST, but in order to evaluate attributes we'd need to do exactly that, since the expression for the attribute value may call functions, depend on parameters, etc.

Reply:
> Hmm.. now it doesn't produce an error message if the expression is not constant.
Yes, attributes with those expressions are now ignored (just like before, when we were ignoring all such attributes) -- I've changed it to only propagate attributes that *are* constant.
This makes it an improvement over what we have in master right now, with nothing lost. 

> Maybe it would make sense to only change the behavior for top attributes?
Of course, the assumption is that `(* top=<constant literal> *)`. We can, of course, whitelist just the `top` attribute if that makes you more comfortable?

Reply:
>  I've changed it to only propagate attributes that are constant.

Yes, I think 6510297712729f9742d00f6a4cd4fc3a0e530758 fixes the issue.

> For module attributes, can it access any parameters?

Icarus Verilog, Verilator, and Verific all seem to accept the following.

```
(* foo = bar + 1 *)
module test #( parameter integer bar = 0 );
endmodule
```